pkg/bucket/client.go

```diff
@@ -0,0 +1,236 @@
+package bucket
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+	"github.com/openshift/oadp-operator/api/v1alpha1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	fileBucketCache = map[types.NamespacedName]string{}
+)
+
+func init() {
+	fileBucketCache = make(map[types.NamespacedName]string)
+}
+
+type Client interface {
+	Exists() (bool, error)
+	Create() (bool, error)
+	Delete() (bool, error)
+	ForceCredentialRefresh() error
+}
+
+func NewClient(b v1alpha1.Bucket, c client.Client) (Client, error) {
+	return &awsBucketClient{bucket: b, client: c}, nil
+}
+
+type awsBucketClient struct {
+	bucket v1alpha1.Bucket
+	client client.Client
+}
+
+var _ Client = &awsBucketClient{}
+
+func (a awsBucketClient) Exists() (bool, error) {
+	s3Client, err := a.getS3Client()
+	if err != nil {
+		return false, err
+	}
+	input := &s3.HeadBucketInput{
+		Bucket: aws.String(a.bucket.Spec.Name),
+	}
+	_, err = s3Client.HeadBucket(input)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			// This is supposed to say "NoSuchBucket", but actually emits "NotFound"
+			// https://github.com/aws/aws-sdk-go/issues/2593
+			case s3.ErrCodeNoSuchBucket, "NotFound":
+				return false, nil
+			default:
+				// Return true, because we are unable to detemine if bucket exists or not
+				return true, fmt.Errorf("unable to determine bucket %v status: %v", a.bucket.Spec.Name, aerr.Error())
+			}
+		} else {
+			// Return true, because we are unable to detemine if bucket exists or not
+			return true, fmt.Errorf("unable to determine bucket %v status: %v", a.bucket.Spec.Name, aerr.Error())
+		}
+	}
+
+	err = a.tagBucket()
+	if err != nil {
+		return true, err
+	}
+
+	return true, nil
+}
+
+func (a awsBucketClient) Create() (bool, error) {
+	s3Client, err := a.getS3Client()
+	if err != nil {
+		return false, err
+	}
+	createBucketInput := &s3.CreateBucketInput{
+		ACL:    aws.String(s3.BucketCannedACLPrivate),
+		Bucket: aws.String(a.bucket.Spec.Name),
+	}
+	if a.bucket.Spec.Region != "us-east-1" {
```

Comment:
Assuming this is because this is the default region?

Reply:
Resolving this thread